### PR TITLE
feat: migrate all CI to inv > tox > util

### LINKED ISSUE #115: ci: use tox for testing
Would allow for the full test suite to be run locally. Specifically:
1. tox checks if there are new dependencies. If so, installs them. This means that if we add a new dependency, but forget to add it to the `.toml`, we'll catch it locally. (I make this mistake too frequently :-()
2. Local testing across python versions, or for a specific python version

Implementaiton-wise, I think we should have: run in the current environment by default, and then:
* `inv pr` - run in selected environments
* `inv test` - run in activated `.venv`. Only run in tox environments if specified (kwarg flag which takes which python versions to run in). 

https://tox.wiki/en/latest/user_guide.html

Should be pretty easy to set up. What do you think? @KennethEnevoldsen :-)

### LINKED ISSUE #124: perf: use tox for github actions
Substantially improves performance, since we can cache the tox environments. Then tox only needs to update any changed dependencies, rather than completely re-create an environment.

Depends on #121 being implemented.

### LINKED ISSUE #125: invoke and github actions: migrate everything to use tox environments
Avoids issues like this one, where we have a dependency that's required for pyright and:

* Installed locally, hence `inv pr` succeeds
* Not installed remotely, hence `pyright .` fails

https://github.com/Aarhus-Psychiatry-Research/timeseriesflattener/actions/runs/4743483039/jobs/8423090479?pr=183

If all the commands use tox environments, we have complete parity.

Do you agree with this prioritisation, @KennethEnevoldsen?

### PR DESCRIPTION
**Pull request by MartinBernstorff**
Closes #115 and closes #125 and closes #124.

### CODE REVIEW

**Comment by MartinBernstorff** [comment]
Would have preferred to use a nice toml parser, but `tomllib` wasn't introduced until python 3.11, and I don't want to introduce more dependencies:

https://docs.python.org/3/library/tomllib.html

File: {{cookiecutter.project_name}}/pyproject.toml [hunk 1]
```diff
@@ -13,7 +13,6 @@ classifiers = [
   "Operating System :: POSIX :: Linux",
   "Operating System :: MacOS :: MacOS X",
   "Operating System :: Microsoft :: Windows",
-  "Programming Language :: Python :: 3",
```
Comment (MartinBernstorff) [inline {{cookiecutter.project_name}}/pyproject.toml:16]: Not sure what this was for. Have removed because it cannot be parsed like the others :-)

**Comment by MartinBernstorff** [comment]
Wait a bit here - it seems tox doesn't automatically update the environment if `pyproject.toml` has changed, kind of defeating the purpose. Will have a look at it later.

**Review by KennethEnevoldsen — COMMENTED** [review]
We need the tests to pass on windows and ubuntu.

**Comment by MartinBernstorff** [comment]
There we go, sorry, didn't notice the failure after I went home :-)

File: {{cookiecutter.project_name}}/pyproject.toml [hunk 2]
```diff
@@ -161,3 +161,16 @@ build_command = "python -m pip install build; python -m build"
 
 [tool.setuptools]
 include-package-data = true
+
+
+[tool.tox]
+legacy_tox_ini = """
+[tox]
+envlist = py{39,310}
```
Comment (KennethEnevoldsen) [inline {{cookiecutter.project_name}}/pyproject.toml:169]: Why don't we use this instead of the tags for denoting python versions (I would expect to change things here not in the tags)

**Review by KennethEnevoldsen — APPROVED** [review]
I would prefer if we could use the envlist instead of the tags but otherwise looks good

File: {{cookiecutter.project_name}}/pyproject.toml (hunk 2 above)
Comment (MartinBernstorff) [reply {{cookiecutter.project_name}}/pyproject.toml:169]: Just to be certain I understand; when you say "tags", I assume you mean e.g. `"Programming Language :: Python :: 3.9"`.

Ideally, I would like only **one** source of truth for supported versions. The closest we can get right now is the tags + the matrix in github actions. 

But I can see how this is currently counterintuitive. I'm not sure how to get those aligned.

Which solution do you think is preferable?

**Comment by MartinBernstorff** [comment]
I've expanded the scope of the PR to migrate everything to the `inv > tox > util` stack 👍

File: {{cookiecutter.project_name}}/pyproject.toml (hunk 2 above)
Comment (KennethEnevoldsen) [reply {{cookiecutter.project_name}}/pyproject.toml:169]: Hmm hard one. 

I would say we need one place to decide on what to test on. Is that the envlist? and one place to denote what the package can run on. Ideally those should be 1-1. Which they mostly are now? (we mostly disregard the envlist?)

File: {{cookiecutter.project_name}}/.github/workflows/tests.yml [hunk 3]
```diff
@@ -30,43 +30,28 @@ jobs:
...
         shell: bash
         run: |
-          # activate venv dependent on the os
-          if [ "${{ matrix.os }}" == "windows-latest" ]; then
-            source .venv/Scripts/activate
-          else
-            source .venv/bin/activate
-          fi
-          pytest --durations=0 -n 2 -x --junitxml=pytest.xml --cov-report=term-missing --cov=src/ tests/
+          inv test --pytest-args="--durations=0" --pytest-args="--junitxml=pytest.xml --cov-report=term-missing --cov=src/ tests/"
```
Comment (KennethEnevoldsen) [inline {{cookiecutter.project_name}}/.github/workflows/tests.yml:54]: two pytest args?

File: {{cookiecutter.project_name}}/tasks.py [hunk 4]
```diff
@@ -317,25 +326,46 @@ def update(c: Context):
     install(c, pip_args="--upgrade", msg=False)
 
 
-@task
-def test(c: Context):
+@task(iterable="pytest_args")
```
Comment (KennethEnevoldsen) [inline {{cookiecutter.project_name}}/tasks.py:329]: Ahh i see - I would probably have it as just a string. No reason to do the parsing.

File: {{cookiecutter.project_name}}/tasks.py [hunk 5]
```diff
@@ -317,25 +326,46 @@ def update(c: Context):
...
+):
     """Run tests"""
+    # Invoke requires lists as type hints, but does not support lists as default arguments.
+    # Hence this super weird type hint and default argument for the python_versions arg.
     echo_header(f"{msg_type.TEST} Running tests")
+
+    python_version_strings = [f"py{v.replace('.', '')}" for v in python_versions]
+    python_version_arg_string = ",".join(python_version_strings)
+
+    if len(pytest_args) == 0:
```
Comment (KennethEnevoldsen) [inline {{cookiecutter.project_name}}/tasks.py:343]:
```suggestion
    if not pytest_args:
```

File: {{cookiecutter.project_name}}/tests/__init__.py
Comment (KennethEnevoldsen) [inline {{cookiecutter.project_name}}/tests/__init__.py:1]: Why is this deleted?

Comment (KennethEnevoldsen) [reply {{cookiecutter.project_name}}/tests/__init__.py:1]: It seems like it is required for pytest to auto find the tests.

File: {{cookiecutter.project_name}}/.github/workflows/tests.yml (hunk 3 above)
Comment (KennethEnevoldsen) [reply {{cookiecutter.project_name}}/.github/workflows/tests.yml:54]:
```suggestion
          inv test --pytest-args="--durations=0 --junitxml=pytest.xml --cov-report=term-missing --cov=src/"
```
It should autofind the tests folder - otherwise you might have the local test succeed and the CI fail.

**Review by KennethEnevoldsen — COMMENTED** [review]
A few minor changes otherwise look good. Frees us from either changing away from invoke or from tox.

File: {{cookiecutter.project_name}}/.github/workflows/tests.yml (hunk 3 above)
Comment (MartinBernstorff) [reply {{cookiecutter.project_name}}/.github/workflows/tests.yml:54]: > It should autofind the tests folder - otherwise you might have the local test succeed and the CI fail.

Agreed, it should not be at the github workflow level.

However, moving it to `inv test`. For moderately sized projects, specifying the tests dir can make test discovery ~1-2 s faster (e.g. `timeseriesflattener`).

File: {{cookiecutter.project_name}}/pyproject.toml (hunk 2 above)
Comment (MartinBernstorff) [reply {{cookiecutter.project_name}}/pyproject.toml:169]: Yeah, we actually need to signal many things:
1. Which envs the package can run on (currently in `pyproject.toml`)
2. Which envs `inv test` runs (currently parses `pyproject.toml`)
3. Which envs `tox` runs by default (currently in the `envlist`)
4. Which envs are run on github actions (currently in `tests.yml`)

Since we basically don't want people running `tox`, I think we can ignore that. Then it's 1:1 with the exception of `test.yml`. There's a trade-off here. 

If we have it run `inv test`, we could parse all the correct environments from the toml. That's nice 👍 
But then we can't use the github actions matrix, so each environment would run sequentially, dramatically increasing test run durations. Not as nice :-(

I think the current situation is pretty good, and really want tox to start moving out to the projects, so merging now. If you come up with a good solution to the above, I'm up for creating the PR 👍

File: {{cookiecutter.project_name}}/.github/workflows/tests.yml (hunk 3 above)
Comment (KennethEnevoldsen) [reply {{cookiecutter.project_name}}/.github/workflows/tests.yml]: I would say 1-2s is not worth it compared to the pain it would be to debug.

> Seems invoke has a bug where two pytest args are required for it to parse them as a list, rather than a string :S

I would probably just parse the string then, this seems like it could cause problems if the user does not know.

File: {{cookiecutter.project_name}}/pyproject.toml (hunk 2 above)
Comment (KennethEnevoldsen) [reply {{cookiecutter.project_name}}/pyproject.toml:169]: That is understandable (that is also why the PR was approved)
